AWS S3를 이용한 프로필 사진 업로드

ahyeonkong opened the pull request:
## #️⃣ Issue Number

#50

## 📝 작업 설명

유저가 본인의 프로필 사진을 업로드하고 조회하는 api를 구현했습니다.
한 장의 프로필 사진을 업로드 할 수 있고, S3에 자동으로 저장됩니다.
프로필 사진 조회 시, URL이 반환됩니다.

프로필 사진 업로드는 여러 번 진행할 수 있기 때문에 수정에 편리한 Patch 메서드를 사용했습니다!

모든 과정은 로그인 후에 진행이 가능합니다. 로그인을 하지 않으면 일단은 프로필 사진 업로드, 프로필 사진 조회가 불가능하도록 구현했습니다. 마이페이지에서 사용하는 api로 생각해주시면 될 것 같습니다.

만약, 다른 유저가 본인의 프로필 조회를 하는 기능이 필요하다면 추가해야 할 것 같습니다.

## 📸스크린샷 (선택)   

## 💬 리뷰 요구사항
일단 최소한으로 동작하도록 구현했고, 리팩토링이 필요할 것 같습니다.


## ✅ PR Checklist

PR이 다음 요구 사항을 충족하는지 확인하세요.

- [x] 커밋 메시지 컨벤션에 맞게 작성했습니다.
- [x] 변경 사항에 대한 테스트를 했습니다.(버그 수정/기능에 대한 테스트).

ahyeonkong committed:
feat: 프로필 사진 업로드 및 조회 구현

- 로그인 시 세션에 사용자 정보 저장
- DB에서 본인 정보 확인 후 프로필 사진 변경, 조회 가능

Resolved: #50

ahyeonkong committed:
feat: 세션에 사용자 이메일 정보 저장

JunBe commented on src/main/java/com/market/saessag/domain/photo/controller/ProfileImageController.java:
```diff
@@ -0,0 +1,68 @@
+                .orElseThrow(() -> new UsernameNotFoundException("사용자를 찾을 수 없습니다."));
+
+        // 프로필 사진 URL 반환
+        return ResponseEntity.ok(UserProfileImageResponse.from(user));
```
s3Service.uploadFile 메서드는 업로드 후 파일명만 반환하게 되어서 user 테이블의 profileUrl에 사진 조회를 할 수 있는 url이 아닌 파일명이 저장이 될텐데 혹시 프로필 사진 조회가 가능한 상태일까요?

ahyeonkong replied:
꼼꼼한 리뷰 감사합니다! 현재는 파일 명만 저장되는 게 맞습니다. 사진 조회를 위해서는 해당 url을 반환하도록 바꾸는 게 맞을 것 같네요!

JunBe replied:
uploadFile 메서드를 파일명만 저장되게 만든 이유가 현재 퍼블릭 엑세스를 차단 시켜둬서 URL로 직접 조회가 불가능하기 때문에,
DB에 저장 되어있는 사진 키(파일 명)를 이용하여 s3Service 내에 있는 getPresignedUrl 메서드에 키를 넣어서 프리사인 URL을 반환하여 프리사인 URL로 사진에 접근하도록 로직을 짠 상태입니다!
해당 URL을 반환하여 사진에 바로 접근하도록 바꾸려면 프리사인 URL을 사용하지 않고, S3에서 퍼블릭 엑세스 차단을 풀고 접근하도록 해야하는데
직접 URL에 접근하는 방식으로 통일하는게 더 편하실까요? 이 방법이 로직은 훨씬 간단해질 것 같습니다!

ahyeonkong replied:
현재 준배님이 작성해주신 getPresignedUrl 메서드를 이용해서 프리사인 URL을 반환하는 코드로 리팩토링 해두었습니다.

프론트엔드에서 조회만 잘 된다면 어떤 방법을 사용하더라도 상관없을 것 같습니다!!

JunBe replied:
넵! 그럼 만약 프론트엔드에서 이 방식으로 조회가 어렵다면 바꾸는 방향으로 가면 될 것 같습니다!

JunBe commented on src/main/java/com/market/saessag/domain/photo/controller/ProfileImageController.java:
```diff
@@ -0,0 +1,68 @@
+
+    // 프로필 사진 업로드
+    @PatchMapping("/image")
+    public ResponseEntity<String> uploadProfileImage(@RequestParam("file") MultipartFile file, HttpSession session) {
```
ApiResponse 사용하기로 한 것 같은데 현재 PhotoController가 ResponseEntity를 사용하고 있어서 사용하신걸까요??
맞다면 PhotoController가 ApiResponse 타입을 반환하도록 리팩토링해도 될까요?

ahyeonkong replied:
넵 그렇습니다! ApiResponse 타입으로 리팩토링 진행하면 좋을 것 같습니다!!

JinTaekLim commented on src/main/java/com/market/saessag/domain/photo/controller/ProfileImageController.java:
```diff
@@ -0,0 +1,68 @@
+    private final UserRepository userRepository;
+
+    // 프로필 사진 업로드
+    @PatchMapping("/image")
```
"/image"만으로는 어떤 API인지 구분하기 어려울 것같습니다. 
해당 코드는 프로필 사진 수정으로 이해했는데 조금 더 정확한 이름과 주석을 달아주시면 좋을 것같아요!

ahyeonkong replied:
넵! 참고해서 수정하겠습니다!

일단, 먼저 설명을 드리면 Patch 메서드 만으로 프로필 사진 등록과 수정을 단일 엔드포인트에서 처리가 가능하다고 생각하여 사용했습니다.
사실 기존 이미지가 없는 상태이고(null), 있다고 하더라도, 결국 사용자 관점에서는 프로필 사진 업로드와 수정이 같은 방식으로 진행되기 때문에 Patch 메서드로 모두 구현이 가능하다고 생각했습니다!

이름과 주석 처리를 자세하게 달아두겠습니다!

JinTaekLim commented on src/main/java/com/market/saessag/domain/photo/controller/ProfileImageController.java:
```diff
@@ -0,0 +1,68 @@
+            // 새 이미지 업로드
+            String fileUrl = s3Service.uploadFile(file);
+            user.setProfileUrl(fileUrl);
+            userRepository.save(user);
```
Repository가 Controller에서 사용되는 건 적절하지 않은 것같다는 생각이 듭니다!
Service단으로 로직 이동 시켜주시면 좋을 것같아요

ahyeonkong replied:
넵!!

JinTaekLim commented on src/main/java/com/market/saessag/domain/photo/dto/UserProfileImageResponse.java:
```diff
@@ -0,0 +1,16 @@
+public class UserProfileImageResponse {
+    private String profileUrl;
+
+    public static UserProfileImageResponse from(User user) {
```
profileUrl 하나의 데이터만 필요한 경우 해당 메소드 없이도 깔끔하게 작성할 수 있을 것같다는 생각이 드는데, 특별히 추가한 이유가 있으실까요?

ahyeonkong replied:
파일명을 반환해야 한다고 이해를 했던 것 같습니다.
그렇기 때문에 해당 메서드를 작성했습니다!
기존 사진 업로드처럼 바로 url을 반환하는 로직으로 변경하면 좋을 것 같습니다!

ahyeonkong committed:
refactor: API 설명 추가 및 프로필 사진 반환 URL 변경

- 프로필 사진 업로드 API를 @PatchMapping(/upload-image)로 변경 및 주석 추가
- 프로필 사진 조회 시 반환하는 URL 변경

ahyeonkong committed:
refactor: 프로필 이미지 관련 비즈니스 로직을 S3Service로 이동

- DI를 통해 UserRepository 의존성을 Controller에서 S3Service로 이동

JinTaekLim commented on src/main/java/com/market/saessag/domain/photo/service/S3Service.java:
```diff
@@ -98,4 +108,27 @@ public Map<String, String> getPresignedUrl(List<String> keys) {
+            throw new CustomException(ErrorCode.PROFILE_IMAGE_NOT_FOUND);
+        }
+
+        return getPresignedUrl(Collections.singletonList(user.getProfileUrl()));
```
getPresignedUrl 에서 [ 파일명 / 주소 ] 형식인 Map을 반환하고 있는 것같은데, 파일명을 활용하고 있는 곳이 있을까요?

ahyeonkong replied:
DB의 profile_url 컬럼에 저장됩니다!

JinTaekLim replied:
현재는 profile_url에 [파일명]이 저장이 되고 필요시에 [파일명]을 이용하여 S3에서 [주소]를 가져오는 형식으로 구현이 되어있는 걸로 이해했는데, profile_url에 [주소]를 저장하지 않고 [파일명]을 저장하게된 의도가 궁금합니다!
@JunBe @ahyeonkong

JunBe replied:
S3 객체에 직접 접근하는 방법은 `https://saessag.s3.ap-northeast-2.amazonaws.com/파일명` 인데 현재 퍼블릭 액세스를 허용하지 않아 접근이 불가능합니다.
현재 접근하는 방법은 DB의 사진 컬럼에 저장 되어있는 파일명을 키로 사용해 프리사인 URL을 생성하여 접속하는 방식입니다.
이때 전체 주소를 저장하면 항상 모든 키에 `https://saessag.s3.ap-northeast-2.amazonaws.com/`가 추가 되기때문에 불필요하다 생각되어 파일명만 저장하게 되었습니다!

ahyeonkong committed:
Merge branch 'develop' of https://github.com/saessagMarket/back into feat/#50-S3-profile-upload

ahyeonkong committed:
refactor: API 응답 형식 표준화

ahyeonkong committed:
refactor: ApiResponse 공통 응답으로 수정

ahyeonkong committed:
refactor: 상태 응답 타입을 int -> HttpStatus로 변경

ahyeonkong committed:
refactor: API 응답 메시지 일관성 개선

- HTTP 상태 코드 문자열을 HttpStatus 객체로 변경
- 응답 메시지 형식 통일

ahyeonkong committed:
feat: SecurityConfig 인증 실패 처리

ahyeonkong committed:
refactor: PathConst에 인증/미인증 API 경로 구분

ahyeonkong committed:
refactor: PathConst 위치 이동으로 구조 개선

ahyeonkong committed:
fix: 엔드포인트의 productId -> id로 수정

ahyeonkong committed:
fix: ProductService의 정렬 에러 해결

- nullLast() 제거
- 정렬 관련 필드명 변경
- INVALID_ARGUMENT 에러 코드 추가

ahyeonkong committed:
fix: 충돌 에러 해결

ahyeonkong committed:
fix: PathConst의 프로필 사진 엔드포인트 수정
